File: C++/2-Medium/00139._单词拆分/00139._单词拆分.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

// 判断字符串s是否能够被拆分为一个或多个在字典wordDict中出现的单词
bool wordBreak(string s, vector<string>& wordDict) {
    unordered_set<string> dict(wordDict.begin(), wordDict.end()); // 将wordDict转换为哈希表，以便快速查询
    vector<bool> dp(s.size()+1, false); // dp[i]表示s的前i个字符是否能被拆分为字典中的单词
    dp[0] = true; // 空字符串可以被拆分

    for(int i = 1; i <= s.size(); ++i){
        for(int j = 0; j < i; ++j){
            if(dp[j] && dict.find(s.substr(j, i-j)) != dict.end()){
                dp[i] = true;
                break;
            }
        }
    }
    return dp[s.size()];
}

int main() {
    // 示例
    vector<string> wordDict = {"leet", "code"};
    string s = "leetcode";
    cout << (wordBreak(s, wordDict) ? "true" : "false") << endl;

    return 0;
}

File: Python/2-Medium/00139._单词拆分/00139._单词拆分.py
from typing import List

def word_break(s: str, wordDict: List[str]) -> bool:
    """
    判断字符串 s 是否可以被空格拆分为 wordDict 中的单词。

    Args:
    s: 待判断的字符串。
    wordDict: 单词字典列表。

    Returns:
    bool: 如果可以拆分，返回 True；否则返回 False。

    示例:
    >>> word_break("leetcode", ["leet", "code"])
    True
    >>> word_break("applepenapple", ["apple", "pen"])
    True
    >>> word_break("catsandog", ["cats", "dog", "sand", "and", "cat"])
    False
    """
    # 将单词字典转化为集合，便于快速查找
    word_set = set(wordDict)
    # 初始化动态规划数组，dp[i] 表示 s[:i] 能否被拆分
    dp = [False] * (len(s) + 1)
    dp[0] = True  # 空字符串可以被拆分
    
    for i in range(1, len(s) + 1):
        for j in range(i):
            # 如果 s[:j] 可以被拆分，并且 s[j:i] 在单词字典中，那么 s[:i] 也可以被拆分
            if dp[j] and s[j:i] in word_set:
                dp[i] = True
                break  # 找到一种拆分方式即可
    
    return dp[len(s)]

# 运行示例测试
test_cases = [
    ("leetcode", ["leet", "code"]),
    ("applepenapple", ["apple", "pen"]),
    ("catsandog", ["cats", "dog", "sand", "and", "cat"])
]

results = [word_break(s, wordDict) for s, wordDict in test_cases]
results
